### ISSUE TITLE
Deep Neural Networks for YouTube Recommendations

### ISSUE THREAD

**Issue by takapy0210** [takapy0210/paper#5]
## Meta
- Link: https://research.google/pubs/pub45530/
- Author: Google
- Date: 
- Journal: Recsys 2016

## どんなもの？（3行ぐらいで）
- Youtubeは次の3点の性質があり、レコメンデーションに課題を抱えていた。
  - スケール、鮮度、ノイズ
- そこで、推薦コンテンツの生成を「候補生成モデル」と「ランキングモデル」に分割し、DNNアルゴリズムを用いることで、従来の行列因数分解アプローチよりも視聴時間が延びた
- また、上記の推薦アーキテクチャを設計、反復、維持することから得られた実践的な教訓と洞察を提供している

## 先行研究と比べてどこがすごい？
- 2段構成のシステムにし、2層のDNNで推薦コンテンツを生成していること
- 後述の2段目のモデルの学習時の目的関数にWeighted Logistic Regressionという、CTRではなく、視聴時間で重み付けした関数を用いることで、重視している"視聴時間"に最適化するようにしたこと

## 技術や手法のキモはどこ？
- データが大規模な文、下記のようの2つのパーツに分けて推薦システムを構築した点
  - 1階層目で対象ユーザーの好みに合いそうな動画を大雑把に絞り込み、2段階目では絞り込まれた動画を精緻にリランキングしている。
  - これらの2つのパーツをDeepLearningで実現した
![image](https://user-images.githubusercontent.com/30361744/96994322-8610e100-1567-11eb-8cb9-ea769af42a02.png)

### 1段階目のモデル
- videoベクトルとuserベクトルからknnで推薦対象のコンテンツ候補を生成しているっぽいけど、videoベクトルはどこからきたのか気になった
- マルチクラス分類の部分が各動画に対する視聴確率みたいなものになってるのかな？
- 特徴量は以下のもの
  - 過去の視聴履歴：視聴動画の埋め込みベクトルを平均化したもの（256次元）
  - 過去の検索履歴：検索内容のunigramとbigramを埋め込んで平均化したもの（256次元）
  - 視聴地域の埋め込みベクトル
  - 訓練サンプルの経過時間（Example Age）と非線形変換（※推論時は0か小さな負数を入力）
  - ユーザーの性別, etc…
![image](https://user-images.githubusercontent.com/30361744/96996257-82328e00-156a-11eb-923b-3328e4f612a2.png)
- ポイントとしては、訓練サンプルの経過時間を入れていることで、コンテンツがどれくらい過去のものかを学習させているところ。
  - Youtubeでは好みと関連が薄くても新鮮なコンテンツが消費されやすい傾向が、事前の分析で分かっていた

### 2段目のモデル
- 使用した特徴量は以下
  - ユーザーに表示した動画の埋め込みベクトル
  - ユーザーが視聴した動画の埋め込みベクトルの平均
  - ユーザー言語と動画言語の埋め込みベクトル
  - 最終視聴からの経過時間と非線形変換
  - 推薦回数, etc…
- 特に、過去に推薦した回数は重要で、何度も推薦しているのにユーザーが反応しないアイテムは推薦から除外する、ということを学習させるためにこの特徴量を入れているみたい
![image](https://user-images.githubusercontent.com/30361744/97017452-e2d0c380-1588-11eb-8917-370ca9c099ea.png)

## どうやって有効だと検証した？
- 定量的な数値は明記されていなかったが、従来よりは精度があがったみたい

## 次に読むべき論文は？
- ニュース推薦に関する論文[17]：[Personalized News Recommendation using Classified Keywords to Capture User Preference](http://icact.org/upload/2014/0412/20140412_finalpaper.pdf)
- 強調フィルタリングとDNN[22]：[Collaborative Deep Learning for Recommender Systems](https://arxiv.org/pdf/1409.2944.pdf)
- 強調フィルタリングとAutoEncoder：[Training Deep AutoEncoders for Collaborative Filtering](https://arxiv.org/pdf/1708.01715.pdf)

## 感想
- コンテンツ候補生成モデルとランキングモデルを分けるのは実務でも活用できそう
- 訓練サンプルの経過時間や過去の推薦した回数などの特徴量は、直感的にも良さそうなので、参考にしたい
- videoベクトルの算出方法がいまいち分からなかったが

## 参考リンク
- [YouTubeの推薦アルゴリズムの変遷を追う〜深層学習から強化学習まで〜](https://medium.com/eureka-engineering/youtube-recommender-algorithm-survey-341a3aa1fbd6)
- https://rf00.hatenablog.com/entry/2018/05/07/211852
- https://tech.dely.jp/entry/youtube_recommend

**Comment by takapy0210** [comment]
## ABSTRACT
YouTubeは、現存するシステムの中で最大規模かつ最も洗練された産業用レコメンデーションシステムの一つである。本論文では、このシステムを高レベルで記述し、ディープラーニングによってもたらされた劇的な性能向上に焦点を当てる。本論文は、古典的な2段階の情報検索の二分法に基づいて構成されている。まず、ディープな候補生成モデルを詳細に説明し、次にディープなランキングモデルを説明する。また、ユーザーに多大な影響を与える大規模なレコメンデーションシステムを設計、反復、維持することから得られた実践的な教訓と洞察を提供する。

候補生成とランキングモデルの2つから構築されている
+
設計、反復、維持することから得られた実践的な教訓と洞察を提供する

**Comment by takapy0210** [comment]
## INTRODUCTION

> YouTubeは、動画コンテンツの作成、共有、発見のための世界最大のプラットフォームです。 YouTubeのレコメンデーションは、10億人以上のユーザーが、増え続ける動画のコーパスからパーソナライズされたコンテンツを発見するのを支援する役割を担っている。 本稿では、ディープラーニングがYouTubeのビデオレコメンデーションシステムに与えたインパクトの大きさに焦点を当てる。  図1は、YouTubeのモバイルアプリのホーム上に表示されている動画の推薦の様子を示している。
YouTubeの動画を推薦することは、以下の3つの観点から非常に困難である。
-スケール：既存の推薦アルゴリズムの多くは、小さな問題でもうまく動作することが証明されていますが、我々のスケールでは動作しません。YouTubeの膨大なユーザーベースとコーパスを扱うためには、高度に特化した分散学習アルゴリズムと効率的な配信システムが不可欠です。
-鮮度：YouTubeは非常にダイナミックなコーパスを持っており、1秒間に何時間もの動画がアップロードされています。  勧告システムは、新たにアップロードされたコンテンツだけでなく、ユーザーが行った最新のアクションをモデル化するために、応答性の高いものでなければなりません。 新しいコンテンツと定着した動画のバランスをとることは、探索・探索の観点からも理解できる。
-ノイズ:YouTube上の過去のユーザー行動は、数が少なく、観察不可能な様々な外的要因があるため、本質的に予測が困難である。  我々はユーザー満足度の根拠となる真実を知ることはほとんどなく、代わりにノイズの多い暗黙のフィードバックシグナルをモデル化しています。  さらに、コンテンツに関連するメタデータは、十分に定義されたオントロジーを持たないまま、構造化されていません。 我々のアルゴリズムは、トレーニングデータのこのような特殊な特性に対してロバストである必要があります。
Googleの他の製品分野との連携により、YouTubeは、ほぼすべての学習問題の汎用的な解決策としてディープラーニングを使用するという根本的なパラダイムシフトを遂げています。 我々のシステムは、最近TensorFlow[1]としてオープンソース化されたGoogleBrain[4]上に構築されています。TensorFlowは、大規模な分散学習を用いて様々なディープニューラルネットワークアーキテクチャを実験するための柔軟なフレームワークを提供します。 我々のモデルは約10億個のパラメータを学習し、数百億個の例題で訓練されている。

Youtubeの推薦は、以下の3つの観点から、難しい問題である

- スケールの問題
既存のレコメンドアルゴリズムの多くは小さな問題ではうまく機能することが証明されているが、YouTubeの規模では機能しない。
YouTubeの規模になるとユーザーデータも莫大にあるからそこをどう対処していくか。

- 新鮮さの問題
毎秒何時間もの大きな動画がアップロードされるのでレコメンドシステムはこれに耐えうる性能でなければならない。
また、ユーザーは新しいコンテンツを好むことはわかっているがそれと引き換えに適合度（関連性）や既存のコンテンツとのバランスは慎重に考慮する必要がある。

- ノイズの問題
YouTubeでの過去のユーザー行動は、スパース性（まばらという意味）と色んな観測不能な外部要因があるから予測がかなり困難。
加えて、各ビデオのメタデータの構造が不十分である。
従って、アルゴリズムを作るにあたってこのような特性を十分に加味しておかなければならない。

**Comment by takapy0210** [comment]
##  CONCLUSIONS

> 我々は、YouTube動画を推薦するためのディープニューラルネットワークアーキテクチャを、候補の生成とランキングという2つの異なる問題に分けて説明した。

> 我々のディープコラボレーションフィルタリングモデルは、多くの信号を効果的に統合し、深さのあるレイヤーとの相互作用をモデル化することができ、YouTube[23]で使用されている以前の行列因数分解アプローチよりも優れています。 推薦のためのサロゲート問題の選択には、より多くの芸術と科学があり、我々は未来の腕時計を分類することで、非対称的な共同腕時計の行動を捉え、未来の情報の漏洩を防ぐことで、ライブメトリクス上で良好なパフォーマンスを発揮することを発見しました。 そうでなければ、モデルはサロゲート問題に適合しすぎてしまい、ホームページにうまく反映されませんでした。

> 我々は、入力特徴としてトレーニングを受けた受験者の年齢を使用することで、過去への固有のバイアスが除去され、モデルが人気のある動画の時間依存の行動を表現できることを実証しました。 これにより、オフラインでのホールドアウト精度が向上し、A/Bテストで最近アップロードされた動画の視聴時間が劇的に増加しました。

> ランキングはより古典的な機械学習の問題であるが、我々のディープラーニングアプローチは、視聴時間予測のための以前の線形およびツリーベースの手法を凌駕していた。 特にレコメンドシステムでは、過去のユーザーの行動を記述した特殊な手法の恩恵を受けることができます。 深層ニューラルネットワークでは、カテゴリカルな特徴と連続的な特徴の特殊な表現を必要とし、それぞれ埋め込みと分位正規化を用いて変換します。  深度の層は、数百の特徴間の非線形相互作用を効果的にモデル化することが示された。

> ロジスティック回帰は、正の例では時計時間、負の例では単数で重み付けすることで修正され、予想される時計時間をほぼモデル化したオッズを学習できるようになった。 このアプローチは、クリックスルー率を直接予測する場合と比較して、ウォッチタイムで重み付けされたランキング評価指標において、はるかに優れたパフォーマンスを示した。

**Comment by takapy0210** [comment]
##  CANDIDATE GENERATION

> 候補の生成中に、巨大なYouTubeのコーパスは、ユーザーに関連する可能性のある数百の動画にまで絞り込まれています。 ここで説明したレコメンダーの前身は、ランクロス[23]で訓練された行列因数分解アプローチであった。 このような観点から、我々のアプローチは、因数分解技術の非線形一般化と見ることができます。

> このような観点から、我々のアプローチは因数分解技術を非線形に一般化したものと見ることができる。

> ここで、u∈RNは、ユーザ、コンテキストペアの高次元エンベッディングを表し、vj∈RNは、各候補動画のエンベッディングを表す。  この設定では、エンベッディングとは、疎なエンティティ（個々の動画、ユーザなど）をRN内の密なベクトルにマッピングしたものである。 YouTubeには明示的なフィードバック機構（サムズアップ/ダウン、製品内アンケートなど）が存在するが、我々は、動画を完成させたユーザーが肯定的な例である時計の非明示的フィードバック[16]をモデルの学習に用いている。 この選択は、暗黙のユーザー履歴の数が桁違いに多いことに基づいており、明示的なフィードバックが非常に疎かになっているテールの奥深くでのレコメンデーションを生成することができます。

> 効率的なエクストリームマルチクラス
> 何百万ものクラスを持つこのようなモデルを効率的に学習するために、バックグランド分布から負のクラスをサンプリングし（「候補サンプリング」）、重要度重み付けによってこのサンプリングを補正する技術が開発されました[10]。各サンプルについて、真のラベルとサンプリングされた負のクラスについて、クロスエントロピーの損失は最小化されます。 実際には、いくつかの数千の負のクラスがサンプリングされ、従来のsoftmaxの100倍以上の高速化に対応しています。 よく知られている変更アプローチに階層型softmax [15]がありますが、我々は同等の精度を達成することができませんでした。  階層型soft-maxでは、ツリーの各ノードをトラバースする際に、関連性のないクラス間の識別が必要となり、分類問題がより困難になり、性能が低下します。

> サービス提供時には、ユーザーに提示するトップNを選択するために、最も可能性の高いNクラス（動画）を計算する必要があります。  数十ミリ秒の厳格なサービス待ち時間の下で何百万ものアイテムをスコアリングするには、クラスの数に従属する近似スコアリングスキームが必要です。 YouTubeの以前のシステムは，ハッシュ処理に依存していましたが，ここで説明する分類器も同様のアプローチを採用しています． softmax出力層からの校正済み尤度は，サービス提供時には不要なので，スコアリング問題は，汎用ライブラリが使用できるドット積空間の最近傍探索に削減されます[12]． A/B再計算結果は、最近傍探索アルゴリズムの選択に特に敏感ではないことがわかりました。

### Model Architecture
> 連続的な単語袋言語モデル[14]に触発されて、我々は、固定語彙の各ビデオの高次元エンベッディングを学習し、これらのエンベッディングをフィードフォワードニューラルネットワークにフィードします。 ユーザの視聴履歴は、スパースな動画IDの可変長シーケンスで表現され、エンベッディングを介して密なベクトル表現にマッピングされる。  ネットワークは固定サイズの密な入力を必要としますが、埋め込みを単純に平均化することで、7つの戦略(sum, component-wise maxなど)の中で最良の結果が得られました。  また、埋め込みは、通常の勾配降下バックプロパゲーション更新により、他のすべてのモデルパラメータと共同で学習されます。 特徴量は広い第一層に連結され、その後、完全に連結された再連結線形単位（ReLU）のいくつかの層が続く [6]。 図3は、後述する非ビデオウォッチ機能を追加した一般的なネットワークアーキテクチャを示しています。

###  Heterogeneous Signals
>  検索履歴はウォッチヒストリーと同様に扱われ、各クエリはユニグラムとビッグラ ムにトークン化され、各トークンは埋め込まれる。 各queryはユニグラムとビッググラムにトークン化され、各to-kenは埋め込まれます。一度平均化されると、ユーザーのトークン化された、em-beddedクエリは、要約された高密度の検索履歴を表します。 ユーザーの性別、ログイン状態、年齢などの単純なバイナリおよび連続的な特徴は、[0,1]に正規化された実数値としてネットワークに直接入力されます。

### Label and Context Selection
> それは、推薦がしばしばasurrogate問題を解決し、特定のコンテキストにtheresultを転送することを強調することが重要です。 古典的な例としては、正確に視聴率を予測することが効果的な映画のレコメンデーションにつながるという仮定があります[2]。 このサロゲート学習問題の選択は、A/Bテストのパフォーマンスに大きな影響を与えるが、オフライン実験では測定が非常に困難であることがわかった。

> 学習例は、我々が生成したレコメンデーションの中の動画だけではなく、YouTubeの全ての動画（他のサイトに埋め込まれている動画も含む）から生成しています。 そうでなければ、新しいコンテンツが表面化することは非常に難しく、レコメンドする側も搾取に偏りすぎてしまうでしょう。ユーザーがレコメンド以外の方法で動画を発見した場合は、共同フィルタリングによって、その発見を他の人に迅速に伝播させることができるようにしたいと考えています。 ライブメトリクスを改善したもう1つの重要な洞察は、ユーザー1人あたりのトレーニング例の数を一定にして、損失関数の中で効果的にユーザーの重みを均等にすることでした。これにより、ユーザーが "tay-lor swift "の検索クエリを発行したばかりの場合の損失ケースを、少数のアクティブなユーザーが支配するのを防ぐことができました。 我々の問題は、次に視聴される動画を予測することであるため、この情報を与えられた分類器は、"tay-lor swift "の検索結果ページに表示される動画が最も視聴される可能性が高いと予測します。 シーケンス情報を捨てて、順序のないトークンの袋で検索結果を表現することで、分類器はラベルの起源を直接認識することができなくなります。

> 動画の自然な消費パターンは、一般的に非対称的な共同視聴確率を発見することにつながります。 エピソードシリーズは、通常、連続して視聴され、ユーザーは、最も広く人気のあるジャンルからアーティストを発見してから、より小さなニッチなジャンルに焦点を当てることがよくあります。 そのため、ユーザーの次の視聴を予測する際には、無作為に保留された視聴を予測するよりも、はるかに優れた性能を発揮することがわかりました（図5）。 多くの協調型フィルタリングシステムは、ランダムなアイテムを持ち出して、ユーザの履歴の他のアイテムからそれを予測することで、ラベルとコンテキストを暗黙的に選択している(5a)。これに対して、我々はユーザの履歴を「ロールバック」して、ランダムな時計を選択し、ユーザがラベル時計をホールドアウトする前に行った行動のみを入力する(5b)。
